packages/next/src/server/web/spec-extension/adapters/request-cookies.ts

```diff
@@ -144,29 +152,86 @@ export class MutableRequestCookiesAdapter {
               )
               try {
                 target.delete(...args)
+                return wrappedCookies
               } finally {
                 updateResponseCookies()
               }
             }
           case 'set':
-            return function (
-              ...args:
-                | [key: string, value: string, cookie?: Partial<ResponseCookie>]
-                | [options: ResponseCookie]
-            ) {
+            return function (...args: SetCookieArgs) {
               modifiedCookies.add(
                 typeof args[0] === 'string' ? args[0] : args[0].name
               )
               try {
-                return target.set(...args)
+                target.set(...args)
+                return wrappedCookies
               } finally {
                 updateResponseCookies()
               }
             }
+
           default:
             return ReflectAdapter.get(target, prop, receiver)
         }
       },
     })
+
+    return wrappedCookies
+  }
+}
+
+export function wrapWithMutableAccessCheck(
```

Comment:
it's not ideal to have to add another proxy. i tried a different approach originally (see the "fix: setting cookies in draftmode" commit) but that felt like it leaks this all over the place (needed a custom `setMutableCookieUnchecked` setter to get around the access check). this feels cleaner